packages/core/lib/commands/call/run.js

```diff
@@ -0,0 +1,167 @@
+module.exports = async function (options) {
+  const fs = require("fs");
+  const util = require("util");
+  const { Environment } = require("@truffle/environment");
+  const OS = require("os");
+  const Codec = require("@truffle/codec");
+  const Encoder = require("@truffle/encoder");
+  const Decoder = require("@truffle/decoder");
+  const TruffleError = require("@truffle/error");
+  const { fetchAndCompile } = require("@truffle/fetch-and-compile");
+  const loadConfig = require("../../loadConfig");
+
+  if (options.url && options.network) {
+    const message =
+      "" +
+      "Mutually exclusive options, --url and --network detected!" +
+      OS.EOL +
+      "Please use either --url or --network and try again." +
+      OS.EOL +
+      "See: https://trufflesuite.com/docs/truffle/reference/truffle-commands/#call" +
+      OS.EOL;
+    throw new TruffleError(message);
+  }
+
+  let config = loadConfig(options);
+  await Environment.detect(config);
+
+  const [contractNameOrAddress, functionNameOrSignature, ...args] = config._;
+  let functionEntry, transaction;
+
+  const { encoder, decoder } = config.fetchExternal
+    ? await sourceFromExternal(contractNameOrAddress, config)
+    : await sourceFromLocal(contractNameOrAddress, config);
+
+  ({ abi: functionEntry, tx: transaction } = await encoder.encodeTransaction(
+    functionNameOrSignature,
+    args
+  ));
+
+  if (functionEntry.stateMutability !== "view") {
+    console.log(
+      `WARNING!!! Not a view function\n  ABI stateMutability: ${functionEntry.stateMutability}`
+    );
+  }
+
+  const provider = new Encoder.ProviderAdapter(config.provider);
+  let result;
+  try {
+    result = await provider.request({
+      method: "eth_call",
+      params: [
+        {
+          from: config.from,
+          to: transaction.to,
+          data: transaction.data
+        },
+        config.blockNumber
+      ]
+    });
+  } catch (error) {
+    console.log(`Error Message: ${error.message}\nError Code: ${error.code}`);
+  }
+
+  if (result !== null && result !== undefined) {
```

Comment:
Oh, wait -- actually, I have that last part wrong.  On looking at this a second time, there's actually some important error-handling missing here.  This is now a blocking comment, sorry, as there's some important stuff not handled here.

So, first off -- the thing I was originally going to write was possibly correct.  An actual *error* from `request`/`sendRequest`, as it's currently written, should be rare, not something expected; as such, we shouldn't be catching and logging such errors, we should just be bubbling them up -- there shouldn't be a `try`/`catch` at all.

But, we *can* get an error back from the client; it just won't be *thrown*.  Rather, it will consist of the return having a different form than you expect.  So that's a case you need to handle.

Now, `request` currently has no handling for that case, so that's something you'll need to write.  Perhaps the right thing to do is to convert this non-thrown error into a thrown one?  In that case, you will need that `try`/`catch` block after all!  But you'll also need quite a bit more handling.

See, in the case that an error is returned from a call, you're going to have to do some work to extract the raw return value so you can decode it.  It's not enough to just print the error message -- that might be OK for a revert string, sure, but what about a panic code?  Or a custom error?  You're going to need the decoder to handle those, and that means you're going to need to get the raw data.

How do you get that raw data?  Well... unfortunately, that varies a bit.  You can see that we have code in Truffle Contract to do that function [here](https://github.com/trufflesuite/truffle/blob/develop/packages/contract/lib/reason.js#L15) -- it may not be the clearest, feel free to bug me if you don't understand what's going on there.  So you'll likely have to do basically that; ideally that functionality should be factored out into some utility, but that may not be practical here, IDK.  Something to discuss with @gnidan, I imagine.

But this is a case that will need to be handled.

Reply:
When implementing the 1193 provider type [here](https://github.com/trufflesuite/truffle/blob/37db8eb29b936443eea59a8feeabfdaf246c6eec/packages/provider/wrapper.js#L11), the error handling has been done when the call response returns an error. With a `try/catch` block, I am able to get the raw data with `error.data` in the `catch` block. I guess decoding that raw data (as you mentioned [can be done like this](https://github.com/trufflesuite/truffle/blob/develop/packages/contract/lib/reason.js#L15)) will solve the cases related with revert strings, panic codes and custom errors.